File: app-geekhome/src/main/java/eu/geekhome/rest/DependencyInjectionBinder.java
package eu.geekhome.rest;

import eu.geekhome.data.Repository;
import eu.geekhome.domain.hardware.HardwareManager;
import eu.geekhome.domain.extensibility.PluginsCoordinator;
import eu.geekhome.domain.automation.AutomationConductor;
import eu.geekhome.domain.automation.BlocklyParser;
import eu.geekhome.domain.automation.blocks.BlockFactoriesCollector;
import eu.geekhome.domain.events.EventsSink;
import eu.geekhome.domain.inbox.Inbox;
import eu.geekhome.rest.inbox.InboxMessageDtoMapper;
import eu.geekhome.rest.live.HeartbeatDtoMapper;
import eu.geekhome.rest.automation.AutomationUnitDtoMapper;
import eu.geekhome.rest.automation.EvaluationResultDtoMapper;
import eu.geekhome.rest.automationhistory.AutomationHistoryDtoMapper;
import eu.geekhome.rest.configurables.ConfigurableDtoMapper;
import eu.geekhome.rest.fields.FieldDefinitionDtoMapper;
import eu.geekhome.rest.hardware.HardwareAdapterDtoMapper;
import eu.geekhome.rest.hardware.NumberedHardwareEventToEventDtoMapper;
import eu.geekhome.rest.hardware.PortDtoMapper;
import eu.geekhome.rest.plugins.PluginDtoMapper;
import eu.geekhome.rest.settinggroup.SettingGroupDtoMapper;
import org.glassfish.hk2.api.Factory;
import org.glassfish.hk2.utilities.binding.AbstractBinder;

import javax.inject.Singleton;

public class DependencyInjectionBinder extends AbstractBinder {

    private final EventsSink _eventsSink;
    private final Repository _repository;
    private final PluginsCoordinator _pluginsCoordinator;
    private final HardwareManager _hardwareManager;
    private final AutomationConductor _automationConductor;
    private final BlockFactoriesCollector _blockFactoriesCollector;
    private Inbox _inbox;

    public DependencyInjectionBinder(
            EventsSink eventsSink,
            Repository repository,
            PluginsCoordinator pluginsCoordinator,
            HardwareManager hardwareManager,
            AutomationConductor automationConductor,
            BlockFactoriesCollector blockFactoriesCollector,
            Inbox inbox
    ) {
        _eventsSink = eventsSink;
        _repository = repository;
        _pluginsCoordinator = pluginsCoordinator;
        _hardwareManager = hardwareManager;
        _automationConductor = automationConductor;
        _blockFactoriesCollector = blockFactoriesCollector;
        _inbox = inbox;
    }

    @Override
    protected void configure() {
        //mappers
        bind(FieldDefinitionDtoMapper.class).to(FieldDefinitionDtoMapper.class).in(Singleton.class);
        bind(ConfigurableDtoMapper.class).to(ConfigurableDtoMapper.class).in(Singleton.class);
        bind(PluginDtoMapper.class).to(PluginDtoMapper.class).in(Singleton.class);
        bind(HardwareAdapterDtoMapper.class).to(HardwareAdapterDtoMapper.class).in(Singleton.class);
        bind(AutomationUnitDtoMapper.class).to(AutomationUnitDtoMapper.class).in(Singleton.class);
        bind(EvaluationResultDtoMapper.class).to(EvaluationResultDtoMapper.class).in(Singleton.class);
        bind(PortDtoMapper.class).to(PortDtoMapper.class).in(Singleton.class);
        bind(NumberedHardwareEventToEventDtoMapper.class).to(NumberedHardwareEventToEventDtoMapper.class).in(Singleton.class);
        bind(AutomationHistoryDtoMapper.class).to(AutomationHistoryDtoMapper.class).in(Singleton.class);
        bind(SettingGroupDtoMapper.class).to(SettingGroupDtoMapper.class).in(Singleton.class);
        bind(HeartbeatDtoMapper.class).to(HeartbeatDtoMapper.class).in(Singleton.class);
        bind(InboxMessageDtoMapper.class).to(InboxMessageDtoMapper.class).in(Singleton.class);

        //factories for objects shared with the App
        bindFactory(new SingletonFactory<>(_hardwareManager)).to(HardwareManager.class).in(Singleton.class);
        bindFactory(new SingletonFactory<>(_automationConductor)).to(AutomationConductor.class).in(Singleton.class);
        bindFactory(new SingletonFactory<>(_pluginsCoordinator)).to(PluginsCoordinator.class).in(Singleton.class);
        bindFactory(new SingletonFactory<>(_repository)).to(Repository.class).in(Singleton.class);
        bindFactory(new SingletonFactory<>(_eventsSink)).to(EventsSink.class).in(Singleton.class);
        bindFactory(new SingletonFactory<>(_blockFactoriesCollector)).to(BlockFactoriesCollector.class).in(Singleton.class);
        bindFactory(new SingletonFactory<>(_inbox)).to(Inbox.class).in(Singleton.class);

        //blocks
        bind(BlocklyParser.class).to(BlocklyParser.class).in(Singleton.class);
    }

    static class SingletonFactory<T> implements Factory<T> {
        private final T instance;

        public SingletonFactory(T instance) {
            this.instance = instance;
        }

        @Override
        public T provide() {
            return instance;
        }

        @Override
        public void dispose(T instance) {
        }
    }
}

File: ae-backend/src/main/java/eu/automateeverything/rest/DependencyInjectionBinder.kt
/*
 * Copyright (c) 2019-2022 Tomasz Babiuk
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  You may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package eu.automateeverything.rest

import eu.automateeverything.data.Repository
import eu.automateeverything.domain.events.EventBus
import eu.automateeverything.domain.extensibility.PluginsCoordinator
import eu.automateeverything.domain.hardware.HardwareManager
import eu.automateeverything.domain.automation.AutomationConductor
import eu.automateeverything.domain.automation.blocks.BlockFactoriesCollector
import eu.automateeverything.domain.inbox.Inbox
import jakarta.inject.Singleton
import eu.automateeverything.domain.automation.BlocklyParser
import eu.automateeverything.domain.dependencies.DependencyChecker
import eu.automateeverything.mappers.*
import org.glassfish.hk2.api.Factory
import org.glassfish.hk2.utilities.binding.AbstractBinder

class DependencyInjectionBinder(
    private val eventBus: EventBus,
    private val repository: Repository,
    private val pluginsCoordinator: PluginsCoordinator,
    private val hardwareManager: HardwareManager,
    private val automationConductor: AutomationConductor,
    private val blockFactoriesCollector: BlockFactoriesCollector,
    private val inbox: Inbox
) : AbstractBinder() {

    override fun configure() {
        //mappers
        bind(FieldDefinitionDtoMapper::class.java)
            .to(FieldDefinitionDtoMapper::class.java)
            .`in`(Singleton::class.java)

        bind(ConfigurableDtoMapper::class.java)
            .to(ConfigurableDtoMapper::class.java)
            .`in`(Singleton::class.java)

        bind(PluginDtoMapper::class.java)
            .to(PluginDtoMapper::class.java)
            .`in`(Singleton::class.java)

        bind(HardwareAdapterDtoMapper::class.java)
            .to(HardwareAdapterDtoMapper::class.java)
            .`in`(Singleton::class.java)

        bind(AutomationUnitDtoMapper::class.java)
            .to(AutomationUnitDtoMapper::class.java)
            .`in`(Singleton::class.java)

        bind(EvaluationResultDtoMapper::class.java)
            .to(EvaluationResultDtoMapper::class.java)
            .`in`(Singleton::class.java)

        bind(PortDtoMapper::class.java)
            .to(PortDtoMapper::class.java)
            .`in`(Singleton::class.java)

        bind(DiscoveryEventMapper::class.java)
            .to(DiscoveryEventMapper::class.java)
            .`in`(Singleton::class.java)

        bind(AutomationHistoryDtoMapper::class.java)
            .to(AutomationHistoryDtoMapper::class.java)
            .`in`(Singleton::class.java)

        bind(SettingGroupDtoMapper::class.java)
            .to(SettingGroupDtoMapper::class.java)
            .`in`(Singleton::class.java)

        bind(HeartbeatDtoMapper::class.java)
            .to(HeartbeatDtoMapper::class.java)
            .`in`(Singleton::class.java)

        bind(InboxMessageDtoMapper::class.java)
            .to(InboxMessageDtoMapper::class.java)
            .`in`(Singleton::class.java)

        bind(LiveEventsMapper::class.java)
            .to(LiveEventsMapper::class.java)
            .`in`(Singleton::class.java)

        bind(BlocklyParser::class.java)
            .to(BlocklyParser::class.java)
            .`in`(Singleton::class.java)

        bind(DependencyChecker::class.java)
            .to(DependencyChecker::class.java)
            .`in`(Singleton::class.java)

        bind(DescriptionsUpdateDtoMapper::class.java)
            .to(DescriptionsUpdateDtoMapper::class.java)
            .`in`(Singleton::class.java)

        //factories for objects shared with the App
        bindFactory(SingletonFactory(hardwareManager))
            .to(HardwareManager::class.java)
            .`in`(Singleton::class.java)

        bindFactory(SingletonFactory(automationConductor))
            .to(AutomationConductor::class.java)
            .`in`(Singleton::class.java)

        bindFactory(SingletonFactory(pluginsCoordinator))
            .to(PluginsCoordinator::class.java)
            .`in`(Singleton::class.java)

        bindFactory(SingletonFactory(repository))
            .to(Repository::class.java)
            .`in`(Singleton::class.java)

        bindFactory(SingletonFactory(eventBus))
            .to(EventBus::class.java)
            .`in`(Singleton::class.java)

        bindFactory(SingletonFactory(blockFactoriesCollector))
            .to(BlockFactoriesCollector::class.java)
            .`in`(Singleton::class.java)

        bindFactory(SingletonFactory(inbox))
            .to(Inbox::class.java)
            .`in`(Singleton::class.java)

    }

    internal class SingletonFactory<T>(private val instance: T) : Factory<T> {
        override fun provide(): T {
            return instance
        }

        override fun dispose(instance: T) {}
    }
}